File: char_count.cpp
#include <iostream>
#include <fstream>

int main() {
    const char* filename = "input.txt";
    const char target = 'a';
    const std::size_t buffer_size = 4096;
    char* buffer = new char[buffer_size];
    int count = 0;

    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        std::cerr << "Failed to open file.\n";
        delete[] buffer;
        return 1;
    }

    while (file) {
        file.read(buffer, buffer_size);
        std::streamsize bytes_read = file.gcount();
        for (std::streamsize i = 0; i < bytes_read; ++i) {
            if (buffer[i] == target) {
                count++;
            }
        }
    }

    delete[] buffer;
    std::cout << "Occurrences of '" << target << "': " << count << std::endl;
    return 0;
}

File: char_count.rs
use std::fs::File;
use std::io::{BufReader, Read};

fn main() {
    let filename = "input.txt";
    let target = b'a';
    let file = File::open(filename).expect("Failed to open file");
    let mut reader = BufReader::new(file);

    let mut buffer = [0; 4096];
    let mut count = 0;

    loop {
        let bytes_read = reader.read(&mut buffer).expect("Failed to read bytes");
        if bytes_read == 0 {
            break;
        }
        count += buffer[..bytes_read]
            .iter()
            .filter(|&&byte| byte == target)
            .count();
    }

    println!("Occurrences of '{}': {}", target as char, count);
}

